src/frontend/src/handler/alter_table.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright 2023 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use anyhow::Context;
+use pgwire::pg_response::{PgResponse, StatementType};
+use risingwave_common::error::{ErrorCode, Result};
+use risingwave_sqlparser::ast::{ColumnDef, ObjectName, Statement};
+use risingwave_sqlparser::parser::Parser;
+
+use super::create_table::{gen_create_table_plan, ColumnIdGenerator};
+use super::privilege::check_super_user;
+use super::{HandlerArgs, RwPgResponse};
+use crate::binder::Relation;
+use crate::{build_graph, Binder, OptimizerContext, TableCatalog};
+
+#[allow(clippy::unused_async)]
+pub async fn handle_add_column(
+    handler_args: HandlerArgs,
+    table_name: ObjectName,
+    new_column: ColumnDef,
+) -> Result<RwPgResponse> {
+    let session = handler_args.session;
+
+    let original_catalog = {
+        let relation = Binder::new(&session).bind_relation_by_name(table_name.clone(), None)?;
+        match relation {
+            Relation::BaseTable(table) if table.table_catalog.is_table() => table.table_catalog,
+            _ => Err(ErrorCode::InvalidInputSyntax(format!(
+                "\"{table_name}\" is not a table or cannot be altered"
+            )))?,
+        }
+    };
+
+    if !check_super_user(&session) && session.user_id() != original_catalog.owner {
+        Err(ErrorCode::PermissionDenied(format!(
+            "must be owner of table \"{table_name}\""
+        )))?
+    }
+
+    // Do not allow altering a table with a connector.
+    if original_catalog.associated_source_id().is_some() {
+        Err(ErrorCode::InvalidInputSyntax(format!(
+            "cannot alter table \"{table_name}\" because it has a connector"
+        )))?
+    }
+
+    // Retrieve the original table definition and parse it to AST.
```

Comment:
For now, we store the SQL and parse it back for `ALTER TABLE` 😄. This works but might be kind of dirty. This can be improved if we 
- refactor the interfaces of `create_table_xxx` and allow direct generating of the stream graph with the given table catalog.
- support derivation definition from the table catalog.